_posts/2020-06-31-SwiftUI-and-TCA-watch-your-perfomance.md

```diff
@@ -0,0 +1,473 @@
+
+```
+struct MyView: View {
+   ....
+    var calculatedProperty: CGFloat {
+        var result: CGFloat = 0
+        // Imagine some heavy operations here
+        // ...
+        return result
+    }
+        
+    init(
+        data: ViewData,
+    ) {
+        self.viewData = data
+        // #1 Transformations done in the view initializer.
+        self.transformedData = data.map { viewData in TransformedData(viewData) }
+        // #2 Calling heavy functions in the initializer.
+        self.calculatedData = performSomeCalculations()
+    }
+ 
+    var body: some View {
+        ChildView(transformedData: transformedData,
+                  calculatedData: calculatedData,
+                  // #3 Everytime body reloads, this calculation is performed.
+                  property: calculateSomethingWhenBodyReloads(),
+                  // #4 The same goes for dynamically calculated properties.
+                  property2: calculatedProperty)
+    }
+    
+    func performSomeCalculations() -> Int {
+        return viewData
+            .enumerated()
+            .map { data in
+                // Some heavy calculations while iterating through the collection.
+                // ...
+            }
+    }
+    
+    func calculateSomethingWhenBodyReloads() -> CGFloat {
+        var result = 0
+        // Another heavy operations may appear here
+        // ...
+        return result
+    }
+}
+
+```
+
+You might think - “I did it multiple times and it works”. Trust me, if you would have more complex views with several child subviews nested and created in the way presented above - sooner or later you will notice performance drops. 
+
+Better option would be to move all of the calculations somewhere else and dispatch to the background thread. The question is - where to move the code from the functions and calculated properties? The answer is **TCA**.
+
+We can declare TCA elements in the following way:
+
+```
+struct MyViewData {
+    public struct State: Equatable {
+        var viewData: [ViewData]
+        var transformedData: [TransformedData]
+        // ...
+        // Other needed data
+    }
+    
+    public enum Action: Equatable {
+        case updateState(with: [ViewData])
+    }
+    
+    public static let reducer = Reducer<State, Action, Void> {
+        state, action, _ in
+        switch action {
+        case let .updateState(viewData):
+            state.transformedData = viewData.map { viewData in TransformedData(viewData) }
+            	// ...
+            	// other heavy calculations moved from MyView.
+            return .none
+        }
+    }
+}
+
+```
+
+Our View looks much more cleaner with the TCA store injected and all things moved out:
+
+```
+struct MyView: View {
+    let store: Store<MyViewData.State, MyViewData.Action>
+        
+    init(store: Store<MyViewData.State, MyViewData.Action>) {
+        self.store = store
+    }
+ 
+    var body: some View {
+        WithViewStore(store) { viewStore in
```

Comment:
`WithViewStore` to jakiś syntax z TCA? Jeśli nie to może `ViewWithStore` będzie bardziej zrozumiałą nazwą

Reply:
Dokładnie, jest to struktura pochodząca z TCA, która pozwala obserwować zmiany w zachodzące w `Store` i odświeżać widoki znajdujące się wewnątrz closure. Także niestety w tym przypadku nazwy nie podmienimy.

Reply:
Dodałem wyjaśnienie z referencją do dokumentacji 😀